### PR TITLE
Audit optional and conditional fields with new tests

### PR DESCRIPTION
**Pull request by bamader**
## Background
We currently have a number of fields marked as either optional or conditional, and we don't have automated verification that these fields behave as expected when they're missing. This PR adds that confirmation by building export functionality for sets of records without optional/conditional fields and then expanding upload tests to include these new data sources.

## GitHub Issue
#1128 

## Validation Plan
Because there was some changing with parsings and validations, I made sure to use the `DevUtility` page to download every version of every file we support, upload them to the tool, and make sure children had the expected number of errors in the expected fields. More spreadsheet uploading and testing of creative cases by any reviewers would be appreciated just to make sure we cover as many upload cases as we can think of. I've added links to download data sets with records missing optional fields and records missing conditional fields to the Dev page for this purpose.

## Automated Testing
- [x] All unit tests are passing.
- [x] All e2e tests are passing.
- [ ] If applicable, unit tests have been added to cover this changeset (not applicable).
- [ ] If applicable, e2e tests have been added to cover this changeset (two new e2e tests were added based off the `upload` model; these tests verify expected behavior of optional fields [which, if they're missing from a record, should cause _no_ errors at all] and conditional fields [which, if they are expected to be present because their conditional is satisfied and they are not, _should_ cause errors]).

## Additional Context
In the process of this PR, I found that we had some inconsistencies in our backend parsing logic that had given rise to a few bugs, so I also tried to correct those here and keep us consistent/robust. They were:
1) `incomeNotDisclosed` wasn't parsing as a boolean because it was set to default to `undefined` rather than false, which means that none of the validations or error modal behavior was catching income issues
2) Our `export` controller wasn't accounting for children with multiple fundings for an enrollment; so in any export we did, if an enrollment had multiple fundings, only the first was appearing as a row
3) We didn't have a validation error for the case where a previous/historical enrollment either didn't' have an exit date, or the exit date was after a child had started a newer enrollment; that was causing wonkiness with conditional tests on `CHANGE_ENROLLMENT` so it's now a requirement that current enrollments must start on or after the exit dates of past enrollments.

### CODE REVIEW

**Review by JordanGuinn — APPROVED** [review]

File: src/controllers/export.ts [hunk 1]
```diff
@@ -26,16 +26,28 @@ export async function streamUploadedChildren(
...
-        // Can just use 0th element of each array as the 'active'/'current'
-        // value because controller.getChildById does presorting for us
-        childStrings.push(
-          flattenChild(columnMetadatas, child, {
-            enrollment,
-            skipInfoForPastEnrollments: i !== 0,
-          })
-        )
-      );
+      child.enrollments?.forEach((enrollment, i) => {
```
Comment (jlhogan) [inline src/controllers/export.ts:29]: prefer `enrollmentIdx` and `fundingIdx` over `i` and `j`

File: src/entity/decorators/Child/enrollmentDatesValidation.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      propertyName: propertyName,
+      options: { message: enrollmentDatesCannotOverlap, ...validationOptions },
+      validator: {
+        async validate(_, { object: child }) {
+          const enrollments = await getManager().find(Enrollment, {
+            where: {
+              childId: (child as Child).id,
+            },
+          });
+          const currentEnrollment = enrollments.find((e) => !e.exit);
```
Comment (jlhogan) [inline src/entity/decorators/Child/enrollmentDatesValidation.ts:25]: make sure this validation does not fail if child has no enrollments nor if enrollments are missing both/either entry and exit dates!

File: src/template/constants.ts [hunk 3]
```diff
@@ -30,7 +30,8 @@ export const REPORTING_PERIOD_FORMATS = [
 ];
 
 // Conditional requirement strings
-export const REQUIRED_IF_US_BORN = 'Required if child was born in the U.S.';
+export const REQUIRED_IF_US_BORN =
+  'Required if the "Birth Certificate Type" field is "US Birth Certificate."';
```
Comment (jlhogan) [inline src/template/constants.ts:34]: why this change?

**Review by jlhogan — APPROVED** [review]

File: client/src/containers/DevUtility/DevUtility.tsx [hunk 4]
```diff
@@ -62,6 +62,36 @@ const DevUtility: React.FC = () => {
             queryParamsAsObject={{ whichFakeChildren: 'missingOne' }}
             downloadText="Download CSV of all children missing one field"
           />
+          <h3>Children Set Missing All Optional Fields</h3>
+          <CSVExcelDownloadButton
+            fileType="xlsx"
+            whichDownload="example"
+            className="margin-bottom-3"
+            queryParamsAsObject={{ whichFakeChildren: 'missingOptional' }}
```
[the remark this answers is not in the document]
Comment (bamader) [reply client/src/containers/DevUtility/DevUtility.tsx:70]: Sure, an enum seems like a natural choice to handle this

File: e2e-tests/tests/uploadMissingOptional.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+const { uploadFile } = require('../utils/uploadFile');
+
+module.exports = {
+  '@tags': ['upload'],
+  uploadMissingOptional: async function (browser) {
+    await browser.init();
+    await browser.timeoutsImplicitWait(10000);
+    await login(browser);
+    await uploadFile(browser, 'missingOptional');
+    // TODO: expect missing info icons in roster
```
[the remark this answers is not in the document]
Comment (bamader) [reply e2e-tests/tests/uploadMissingOptional.js:11]: That, my friend, is a piece of kruft from an age long passed (more specifically, before we were just pulling the number of children with missing info from the batch edit banner alert--the comment is superfluous now)

File: src/template/constants.ts (hunk 3 above)
Comment (bamader) [reply src/template/constants.ts:34]: To me it was just a slightly clearer choice of language. I thought there could be parents with children born in the US who may not have the birth certificate available, and so would still try to pick US Birth Certificate (which would then throw errors). That may be desired, though, and if this change doesn't actually clarify anything I can switch it back.

File: src/entity/decorators/Child/enrollmentDatesValidation.ts (hunk 2 above)
Comment (bamader) [reply src/entity/decorators/Child/enrollmentDatesValidation.ts:25]: Ooooo good catch, since we've def had issues with validations crashing like that before
